src/tests/CullingTest.cpp

```diff
@@ -0,0 +1,210 @@
+
+  std::string sceneFile =
+      Cr::Utility::Directory::join(TEST_ASSETS, "objects/5boxes.glb");
+
+  // load the scene
+  int sceneID = sceneManager.initSceneGraph();
+  auto& sceneGraph = sceneManager.getSceneGraph(sceneID);
+  esp::scene::SceneNode& sceneRootNode = sceneGraph.getRootNode();
+  auto& drawables = sceneGraph.getDrawables();
+  const esp::assets::AssetInfo info =
+      esp::assets::AssetInfo::fromPath(sceneFile);
+  bool loadSuccess =
+      resourceManager.loadScene(info, &sceneRootNode, &drawables);
+  EXPECT_EQ(loadSuccess, true);
+
+  // set the camera
+  esp::gfx::RenderCamera& renderCamera = sceneGraph.getDefaultRenderCamera();
+
+  // The camera to be set:
+  // pos: {7.3589f, -6.9258f,4.9583f}
+  // rotation: 77.4 deg, around {0.773, 0.334, 0.539}
+  // fov = 39.6 deg
+  // resolution: 800 x 600
+  // clip planes (near: 0.1m, far: 100m)
+  // with such a camera, the box 3 should be invisible, box 0, 1, 2, 4 should be
+  // visible.
+
+  // NOTE: the following test reults have been visually verified in utility
+  // viewer
+  renderCamera.setProjectionMatrix(800,     // width
+                                   600,     // height
+                                   0.01f,   // znear
+                                   100.0f,  // zfar
+                                   39.6f);  // hfov
+
+  esp::scene::SceneNode agentNode = sceneGraph.getRootNode().createChild();
+  esp::scene::SceneNode cameraNode = agentNode.createChild();
+  cameraNode.translate({7.3589f, -6.9258f, 4.9583f});
+  const Mn::Vector3 axis{0.773, 0.334, 0.539};
+  cameraNode.rotate(Mn::Math::Deg<float>(77.4f), axis.normalized());
+  renderCamera.node().setTransformation(cameraNode.absoluteTransformation());
+
+  // collect all the drawables and their transformations
+  std::vector<std::pair<std::reference_wrapper<Mn::SceneGraph::Drawable3D>,
+                        Mn::Matrix4>>
+      drawableTransforms = renderCamera.drawableTransformations(drawables);
+
+  // do the culling (to create the testing group)
+  std::vector<std::pair<std::reference_wrapper<Magnum::SceneGraph::Drawable3D>,
+                        Magnum::Matrix4>>::iterator newEndIter =
+      renderCamera.cull(drawableTransforms);
+
+  // ============== Test 1 ==================
+  // draw all the invisibles reported by cull()
+  // check passed if 0 sample is returned (that means they are indeed
+  // invisibles.)
+  {
+    std::vector<std::pair<std::reference_wrapper<Mn::SceneGraph::Drawable3D>,
+                          Mn::Matrix4>>
+        objects;
+    for_each(
+        newEndIter, drawableTransforms.end(),
+        [&](const std::pair<std::reference_wrapper<Mn::SceneGraph::Drawable3D>,
+                            Mn::Matrix4>& a) { objects.emplace_back(a); });
+    Mn::GL::SampleQuery q{Mn::GL::SampleQuery::Target::AnySamplesPassed};
+    q.begin();
+    renderCamera.MagnumCamera::draw(objects);
+    q.end();
+
+    EXPECT_EQ(q.result<bool>(), false);
```

Comment:
It is funny the check would fail here indicating that there is something visible. So I hacked the viewer. 
It turns out the "newEndIter" contains 1 actually visible object. I am really wondering  if "remove_if" indeed moved the visible objects to the first half. I am still debugging it.

Reply:
Update: It is solved. See the current code.
The invisibles are not stored in the 2nd part of the vector because std::remove_if only move elements, not swap elements.

Reply:
Argh, sorry, for some reason I thought `std::remove_if()` non-destructively partitions the elements. [std::partition()](https://en.cppreference.com/w/cpp/algorithm/partition) is what should be used there, and then it returns both parts correctly. With that I think the `object.erase()` block above isn't needed anymore either, right?

Another option is `std::stable_partition()` but I don't think we need to preserve the order.

Reply:
if remove_if() is used, then object.erase() is needed if I understand the docs correctly.
https://en.cppreference.com/w/cpp/algorithm/remove

Reply:
Sure, sure, I meant regarding the 

> std::remove_if only move elements, not swap elements

Then, with `std::partition()` the second call to `drawableTransforms()` and the complex block after in the test with `std::addressof()` inside is no longer needed and it could be just:

```cpp
std::vector objectsVisible{drawableTransforms.begin(), newEndIter};
std::vector objectsCulled{newEndIter, drawableTransforms.end()};
```

and then continuing on each of those.

Reply:
Hi @mosra,
Yes, `std::partition` can divide the set into two sub-sets, but it is slower than `std::remove_if`. 
There is just no "free lunch". This piece of code is performance-critical. So I would like to keep `std::remove_if`.
By using `std::remove_if`, the entries after the new end will still have their old values. It is fine, as we are only bothered by this fact in the tests.

Reply:
Is the difference actually measurable? Don't think so. AFAICT the difference is that one does `std::swap(a, b)` while the other `a = std::move(b)`, and I'd say the amount of operations is roughly the same, as `std::partition` doesn't need to keep the order.

Reply:
I wrote a test as follows. Hope it is correct.

```cpp
#include <algorithm>
#include <iostream>
#include <sys/time.h>
#include <time.h>
#include <vector>

double cpuSecond() {
  struct timeval tp;
  gettimeofday(&tp, NULL);
  return ((double)tp.tv_sec + (double)tp.tv_usec * 1.e-6);
}

struct matrix4 {
  double e[16];
  matrix4() {
    for (int i = 0; i < 16; ++i) {
      e[i] = (double)(rand() / 10000) / 100.00;
    }
  }
};

const int range = 10000;

bool isInFirstHalf(const std::pair<int, matrix4> &a) {
  if (a.first < range / 2) {
    return true;
  }
  return false;
};

int main() {
  int nElem = 1 << 26;
  srand(time(NULL));

  auto testFunc =
      [&](std::vector<std::pair<int, matrix4>>::iterator (*func)(
              std::vector<std::pair<int, matrix4>>::iterator beginIter,
              std::vector<std::pair<int, matrix4>>::iterator endIter,
              bool (*criterion)(const std::pair<int, matrix4> &a)),
          const std::string &funcNameStr, int rounds) {
        double totalTime = 0.0;
        std::cerr << std::endl
                  << " ====== Testing: " << funcNameStr
                  << " ======" << std::endl;
        for (int iRound = 0; iRound < rounds; ++iRound) {
          std::cerr << "Round " << iRound << " : "
                    << "Preparing data... ";
          std::vector<std::pair<int, matrix4>> testVector;
          for (int i = 0; i < nElem; ++i) {
            testVector.emplace_back(rand() / range, matrix4{});
          }
          std::cerr << "DONE. ";
          double start = cpuSecond();
          (*func)(testVector.begin(), testVector.end(), isInFirstHalf);
          double elaps = cpuSecond() - start;
          std::cerr << " Time used by " << funcNameStr << ": " << elaps
                    << " sec." << std::endl;

          totalTime += elaps;
        }
        std::cerr << "Total time used by " << funcNameStr << ": " << totalTime
                  << " sec." << std::endl;
        std::cerr
            << "NOTE: Time to prepare data is excluded from above results."
            << std::endl;
      };


  const int rounds = 5;

  testFunc(&std::remove_if, "std::remove_if", rounds);
  testFunc(&std::partition, "std::partition", rounds);

  return 0;
}
```

Result is as follows:
```
 ====== Testing: std::remove_if ======
Round 0 : Preparing data... DONE.  Time used by std::remove_if: 1.78626 sec.
Round 1 : Preparing data... DONE.  Time used by std::remove_if: 1.78558 sec.
Round 2 : Preparing data... DONE.  Time used by std::remove_if: 1.78178 sec.
Round 3 : Preparing data... DONE.  Time used by std::remove_if: 1.77711 sec.
Round 4 : Preparing data... DONE.  Time used by std::remove_if: 1.78206 sec.
Total time used by std::remove_if: 8.91279 sec.
NOTE: Time to prepare data is excluded from above results.

 ====== Testing: std::partition ======
Round 0 : Preparing data... DONE.  Time used by std::partition: 1.84218 sec.
Round 1 : Preparing data... DONE.  Time used by std::partition: 1.84741 sec.
Round 2 : Preparing data... DONE.  Time used by std::partition: 1.85215 sec.
Round 3 : Preparing data... DONE.  Time used by std::partition: 1.85704 sec.
Round 4 : Preparing data... DONE.  Time used by std::partition: 1.85889 sec.
Total time used by std::partition: 9.25767 sec.
NOTE: Time to prepare data is excluded from above results.

```

I also switched the testing sequence (testing std::partition first).
Result is as follows: 
```
 ====== Testing: std::partition ======                                                                                
Round 0 : Preparing data... DONE.  Time used by std::partition: 1.86608 sec.                                          
Round 1 : Preparing data... DONE.  Time used by std::partition: 1.86733 sec.                                          
Round 2 : Preparing data... DONE.  Time used by std::partition: 1.87532 sec.                                          
Round 3 : Preparing data... DONE.  Time used by std::partition: 1.88619 sec.                                          
Round 4 : Preparing data... DONE.  Time used by std::partition: 1.8821 sec.                                           
Total time used by std::partition: 9.37702 sec.                                                                       
NOTE: Time to prepare data is excluded from above results.                                                            
                                                                                                                      
 ====== Testing: std::remove_if ======                                                                                
Round 0 : Preparing data... DONE.  Time used by std::remove_if: 1.82084 sec.                                          
Round 1 : Preparing data... DONE.  Time used by std::remove_if: 1.80141 sec.                                          
Round 2 : Preparing data... DONE.  Time used by std::remove_if: 1.80309 sec.                                          
Round 3 : Preparing data... DONE.  Time used by std::remove_if: 1.79911 sec.                                          
Round 4 : Preparing data... DONE.  Time used by std::remove_if: 1.79908 sec.                                          
Total time used by std::remove_if: 9.02353 sec.                                                                       
NOTE: Time to prepare data is excluded from above results.   
```

Conclusion: `std::remove_if` is slightly faster than `std::partition`.

Reply:
Intuitively, `swap(a, b)` involves 3 move assignments while `a = std::move(b)` is just one.

Reply:
I wrote a test. Hope it is correct and helps.
```cpp
#include <algorithm>
#include <iostream>
#include <sys/time.h>
#include <time.h>
#include <vector>

double cpuSecond() {
  struct timeval tp;
  gettimeofday(&tp, NULL);
  return ((double)tp.tv_sec + (double)tp.tv_usec * 1.e-6);
}

struct matrix4 {
  double e[16];
  matrix4() {
    for (int i = 0; i < 16; ++i) {
      e[i] = (double)(rand() / 10000) / 100.00;
    }
  }
};

const int range = 10000;

bool isInFirstHalf(const std::pair<int, matrix4> &a) {
  if (a.first < range / 2) {
    return true;
  }
  return false;
};

int main() {
  int nElem = 1 << 26;
  srand(time(NULL));

  auto testFunc =
      [&](std::vector<std::pair<int, matrix4>>::iterator (*func)(
              std::vector<std::pair<int, matrix4>>::iterator beginIter,
              std::vector<std::pair<int, matrix4>>::iterator endIter,
              bool (*criterion)(const std::pair<int, matrix4> &a)),
          const std::string &funcNameStr, int rounds) {
        double totalTime = 0.0;
        std::cerr << std::endl
                  << " ====== Testing: " << funcNameStr
                  << " ======" << std::endl;
        for (int iRound = 0; iRound < rounds; ++iRound) {
          std::cerr << "Round " << iRound << " : "
                    << "Preparing data... ";
          std::vector<std::pair<int, matrix4>> testVector;
          for (int i = 0; i < nElem; ++i) {
            testVector.emplace_back(rand() % range, matrix4{});
          }
          std::cerr << "DONE. ";
          double start = cpuSecond();
          (*func)(testVector.begin(), testVector.end(), isInFirstHalf);
          double elaps = cpuSecond() - start;
          std::cerr << " Time used by " << funcNameStr << ": " << elaps
                    << " sec." << std::endl;

          totalTime += elaps;
        }
        std::cerr << "Total time used by " << funcNameStr << ": " << totalTime
                  << " sec." << std::endl;
        std::cerr
            << "NOTE: Time to prepare data is excluded from above results."
            << std::endl;
      };


  const int rounds = 5;

  testFunc(&std::remove_if, "std::remove_if", rounds);
  testFunc(&std::partition, "std::partition", rounds);

  return 0;
}
```

The result is as follows:
```
 ====== Testing: std::remove_if ======
Round 0 : Preparing data... DONE.  Time used by std::remove_if: 1.74562 sec.
Round 1 : Preparing data... DONE.  Time used by std::remove_if: 1.73391 sec.
Round 2 : Preparing data... DONE.  Time used by std::remove_if: 1.73497 sec.
Round 3 : Preparing data... DONE.  Time used by std::remove_if: 1.73419 sec.
Round 4 : Preparing data... DONE.  Time used by std::remove_if: 1.73396 sec.
Total time used by std::remove_if: 8.68264 sec.
NOTE: Time to prepare data is excluded from above results.

 ====== Testing: std::partition ======
Round 0 : Preparing data... DONE.  Time used by std::partition: 1.75068 sec.
Round 1 : Preparing data... DONE.  Time used by std::partition: 1.76395 sec.
Round 2 : Preparing data... DONE.  Time used by std::partition: 1.76781 sec.
Round 3 : Preparing data... DONE.  Time used by std::partition: 1.80863 sec.
Round 4 : Preparing data... DONE.  Time used by std::partition: 1.75734 sec.
Total time used by std::partition: 8.84842 sec.
NOTE: Time to prepare data is excluded from above results.

```

I also changed the testing sequence (test `std::partition` first).
The result is as follows:
```
 ====== Testing: std::partition ======
Round 0 : Preparing data... DONE.  Time used by std::partition: 1.77457 sec.
Round 1 : Preparing data... DONE.  Time used by std::partition: 1.77498 sec.
Round 2 : Preparing data... DONE.  Time used by std::partition: 1.77569 sec.
Round 3 : Preparing data... DONE.  Time used by std::partition: 1.77508 sec.
Round 4 : Preparing data... DONE.  Time used by std::partition: 1.77671 sec.
Total time used by std::partition: 8.87702 sec.
NOTE: Time to prepare data is excluded from above results.

 ====== Testing: std::remove_if ======
Round 0 : Preparing data... DONE.  Time used by std::remove_if: 1.73139 sec.
Round 1 : Preparing data... DONE.  Time used by std::remove_if: 1.73032 sec.
Round 2 : Preparing data... DONE.  Time used by std::remove_if: 1.7295 sec.
Round 3 : Preparing data... DONE.  Time used by std::remove_if: 1.72571 sec.
Round 4 : Preparing data... DONE.  Time used by std::remove_if: 1.73221 sec.
Total time used by std::remove_if: 8.64913 sec.
NOTE: Time to prepare data is excluded from above results.
```

Conclusion: `std::remove_if` is slightly faster than `std::partition`.

Reply:
Update: I improved the above program tonight in order to do apple-by-apple comparison, namely, both methods work on exactly the same testing data.

The conclusion did not change. `std::remove_if` is a bit faster.

Reply:
Fair enough. Pick whatever you prefer, I'd say the difference is not so huge so i'd still choose convenience in the test.